Add col bcast for where binary variant TTS and TST

Aswinmcw opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/27624

### Problem description
where op llk does not have col broadcast support

### What's changed
provide native bcast support for TTS and TST variant

For 32x32: ~56% improved
LLK version - 6159.00 ns
Legacy version - 13953.00 ns

For 1024x1024: ~80% improved
LLK version - 45608.00 ns
Legacy version - 230796.00 ns

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
https://github.com/tenstorrent/tt-metal/actions/runs/17673906164
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
https://github.com/tenstorrent/tt-metal/actions/runs/17673909619
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-quick-trigger.yaml) CI passes (if applicable)
- [ ] [TG demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

Aswinmcw committed:
Add col bcast for where binary variant

Aswinmcw committed:
Fix tensor shape in where operation test cases and adjust true_tile_offset logic in the kernel for column broadcast handling.

Aswinmcw committed:
Update where operation tests and improve logging in the kernel

### Changes made
- Re-enable a test case for the where operation in `test_where.py` and adjust parameterization for variants and scalars.
- Change logging from `log_info` to `log_debug` in the `where.cpp` kernel to reduce verbosity.
- Simplify shape calculations in `where_program_factory.cpp` by removing ceiling division for height and width.
- Clean up comments and improve clarity in the `where_utils.cpp` and `ternary_reader_colbcast_tts.cpp` files.

### Impact
These changes enhance test coverage and improve logging clarity, contributing to better debugging and maintenance of the where operation functionality.

Aswinmcw committed:
Enhance where operation tests and improve broadcast validation logic

### Changes made
- Re-enable and expand test cases in `test_where.py` to cover additional scenarios for the where operation.
- Update parameterization for variants and scalars in the tests.
- Add detailed validation for COL_BCAST and ROW_BCAST cases in `where_device_operation.cpp`.
- Improve broadcasting logic in `where_program_factory.cpp` and `where_utils.cpp` to handle multi-dimensional broadcasting and ensure compatibility checks.

### Impact
These changes improve test coverage and enhance the robustness of the where operation's broadcasting functionality, facilitating better handling of edge cases.

Aswinmcw committed:
Refactor where operation tests and simplify broadcast validation logic

### Changes made
- Remove outdated test case for TTS column broadcast from `test_where.py`.
- Streamline broadcast validation logic in `where_device_operation.cpp` by eliminating specific height and width checks for COL_BCAST and ROW_BCAST cases.

### Impact
These changes enhance the clarity and maintainability of the where operation's test suite and validation logic, ensuring better alignment with broadcasting requirements.

Aswinmcw committed:
Merge main branch with outer broadcast support into column broadcast feature

- Added support for both outer broadcast and column broadcast in TTS where operation
- Preserved existing column broadcast implementation while integrating new outer broadcast kernels
- Resolved conflicts to ensure both broadcast types work correctly
- Updated test cases to include tests for both broadcast types

Aswinmcw committed:
fix

Aswinmcw committed:
fix

Aswinmcw committed:
fix

Aswinmcw committed:
fix

Aswinmcw committed:
fix

Aswinmcw committed:
rename

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/where_binary_col_bcast

KalaivaniMCW commented on tests/ttnn/unit_tests/operations/eltwise/test_where.py:
```diff
@@ -78,24 +78,43 @@ def test_ttnn_where(c_shape, t_shape, f_shape, scalar, variant, condition, devic
         ((2, 2, 1024, 1024), (1, 1024), (1024, 1024)),
         # Bcast cases for dim -1 (col)
         ((1024), (1), (1024)),
+        ((1, 1, 1024, 1024), (1, 1, 1024, 1), (1, 1, 1024, 1024)),  # A, Bcol, C
```
You can just add these shapes to the above test case. 
In that case, You wouldn't have to modify this one below

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/where_binary_col_bcast

Aswinmcw committed:
simplify

Aswinmcw committed:
Enhance `test_where.py` with additional test cases for broadcasting scenarios and simplify the `test_ttnn_where_bcast` function by removing variant parameter handling. Update `where_program_factory.cpp` to streamline shape dimension extraction for value_true tensor.

Aswinmcw committed:
Update `where_program_factory.cpp` to increase the number of kernel arguments from 3 to 4, enhancing the functionality of the runtime argument setup.

Aswinmcw committed:
Remove unnecessary comments related to predicate dimensions in `tts_tst_reader_col_bcast.cpp` to streamline the code and improve readability.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/where_binary_col_bcast

Aswinmcw committed:
Refactor get_broadcast_type function to remove variant parameter across multiple files for improved clarity and consistency in broadcast detection logic.

Aswinmcw committed:
Refactor get_broadcast_type function to remove the variant parameter in multiple files, enhancing clarity and consistency in broadcast detection logic.

Aswinmcw committed:
Remove outdated comments regarding outer broadcast validation in WhereDeviceOperation, improving code clarity.

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/ternary/where/device/where_device_operation.cpp:
```diff
@@ -225,6 +243,57 @@ TensorSpec WhereDeviceOperation::compute_output_specs(
         return ttnn::Shape(output_shape);
     };
 
+    const auto compute_broadcasted_output_tts = [&]() {
```
is compute_broadcasted_output_binary missing any cases ? 
why do we need another one here  ?

Aswinmcw committed:
Refactor WhereDeviceOperation to streamline broadcasting logic by removing the compute_broadcasted_output_tts function and consolidating TTS handling into a single binary function, enhancing code clarity and maintainability.

Aswinmcw committed:
Refactor WhereDeviceOperation to simplify broadcasting logic by removing unnecessary condition related to TTS variant, enhancing code clarity and maintainability.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/where_binary_col_bcast

Aswinmcw committed:
Enhance validation logic in WhereDeviceOperation for TST and TTS variants by ensuring identical heights for column broadcast. This improves clarity and correctness in broadcasting requirements.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/where_binary_col_bcast

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/ternary/where/device/where_device_operation.cpp:
```diff
@@ -84,22 +85,48 @@ void WhereDeviceOperation::validate_on_program_cache_miss(
+            const auto pred_h = predicate_tensor.logical_shape()[-2];
+            const auto true_h = value_true_tensor.value().logical_shape()[-2];
+            TT_FATAL(
+                pred_h == true_h,
```
If you're checking subtile criteria, pls add the check for w-dim as well. 
This doesn't check the W-dim
If your intention is to validate the subtile bcast criteria, you need to include both H and W

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/ternary/where/device/where_program_factory.cpp:
```diff
@@ -150,23 +150,40 @@ void set_or_update_runtime_arguments(
+            }
+
+            // Initialize dimensions for TTS
+            const auto out_rank = output.logical_shape().rank();
+            aND = extract_nD_dims(predicate_tensor, out_rank);           // predicate nD
```
unnecessary. Is already computed and available in a higher scope.

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/ternary/where/device/where_program_factory.cpp:
```diff
@@ -150,23 +150,40 @@ void set_or_update_runtime_arguments(
+            const auto out_rank = output.logical_shape().rank();
+            aND = extract_nD_dims(predicate_tensor, out_rank);           // predicate nD
             bND = extract_nD_dims(value_true_tensor.value(), out_rank);  // value_true nD
+            cND = extract_nD_dims(output, out_rank);                     // output nD
```
unnecessary. Is already computed and available in a higher scope.

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/ternary/where/device/where_program_factory.cpp:
```diff
@@ -179,7 +196,7 @@ void set_or_update_runtime_arguments(
 
             // Extended broadcast arguments
-            if (broadcast_type == WhereBroadcastType::OUTER_BCAST) {
+            if (broadcast_type == WhereBroadcastType::OUTER_BCAST || broadcast_type == WhereBroadcastType::COL_BCAST) {
```
if (broadcast_type != WhereBroadcastType::NONE)

Aswinmcw committed:
Refactor where_program_factory and where_utils to improve broadcasting logic by removing unnecessary dimension extraction and adding validation for invalid broadcast scenarios. This enhances clarity and correctness in handling tensor shapes.

Aswinmcw committed:
Refactor where_program_factory to streamline dimension extraction for TTS, removing unnecessary variable declarations and enhancing code clarity.

Aswinmcw committed:
Refactor where_program_factory to improve broadcasting argument handling by removing unnecessary variable declarations and simplifying the broadcast type check. This enhances code clarity and correctness in runtime argument setup.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/where_binary_col_bcast

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/ternary/where/device/where_program_factory.cpp:
```diff
@@ -337,13 +349,85 @@ void set_or_update_runtime_arguments(
+
+            if (broadcast_type == WhereBroadcastType::COL_BCAST) {
+                // For TTS column broadcast, calculate freq and counter for dedicated kernel
+                const auto& output_shape = output.padded_shape();
+                const auto& tile = output.tensor_spec().tile();
+                uint32_t output_Ht = output_shape[-2] / tile.get_height();
+                uint32_t output_Wt = output_shape[-1] / tile.get_width();
```
duplicated ? 
pls check carefully.

Aswinmcw committed:
fix rebase

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/ternary/where/where.cpp:
```diff
@@ -189,7 +191,7 @@ Tensor WhereOperation::invoke(
     }
 
-    log_debug(tt::LogOp, "Where - legacy");
+    log_info(tt::LogOp, "Where - legacy");
```
use log_info here only for local testing

Aswinmcw replied:
yes will be replaced once testing is done

Aswinmcw committed:
Enhance validation for column broadcast in WhereDeviceOperation

- Updated validation logic to ensure that for column broadcast, the heights of the predicate and true/false value tensors are identical.
- Added checks to confirm that exactly one of the widths is 1, improving error handling and clarity in the Where TTS and TST operations.
- Refactored calculations for frequency and counter in set_or_update_runtime_arguments to use computed height and width instead of output shape, ensuring consistency across broadcast operations.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/where_binary_col_bcast

Aswinmcw committed:
update

Aswinmcw committed:
Refactor broadcast logic in where_program_factory.cpp to simplify frequency and counter calculations. Removed unnecessary height and width broadcast checks, streamlining the runtime argument setup for dedicated kernels.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/where_binary_col_bcast

Aswinmcw committed:
Update submodule references and change log level from info to debug in WhereOperation class

Aswinmcw committed:
Fix variable naming in where_program_factory.cpp to improve clarity and correctness of tile calculations

Aswinmcw committed:
Enhance COL_BCAST handling in where_program_factory.cpp by adding sharding checks and updating runtime arguments for improved tensor broadcasting support.

Aswinmcw committed:
Refactor broadcast handling in tts_tst_reader_col_bcast.cpp to unify true and false tensor broadcasting checks under SRC_BCAST_CB1 macro. This change enhances clarity and maintains consistency in offset calculations for tensor operations.

Aswinmcw committed:
Enhance test coverage in test_where.py by adding a new case for tensor shape variations. Additionally, refactor where_program_factory.cpp to correct the order of tensor data types in the dataflow defines for improved clarity and consistency.

Aswinmcw committed:
Refactor broadcast argument handling in where_program_factory.cpp to simplify conditions for runtime arguments. This change enhances clarity by removing unnecessary checks for OUTER_BCAST and consolidating logic under a single condition for non-NONE broadcast types.

Aswinmcw committed:
Update ttnn_basic_operations.py

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/where_binary_col_bcast
